src/main/drivers/compass/compass_ak8963.c

```diff
@@ -255,48 +343,114 @@ static bool ak8963Read(int16_t *magData)
     magData[Y] = -(int16_t)(buf[3] << 8 | buf[2]) * magGain[Y];
     magData[Z] = -(int16_t)(buf[5] << 8 | buf[4]) * magGain[Z];
 
-#if defined(MPU6500_SPI_INSTANCE) || defined(MPU9250_SPI_INSTANCE)
-    state = CHECK_STATUS;
-#endif
-    return ak8963SensorWrite(AK8963_MAG_I2C_ADDRESS, AK8963_MAG_REG_CNTL1, CNTL1_MODE_ONCE); // start reading again
+    return ak8963WriteRegister(busdev, AK8963_MAG_REG_CNTL1, CNTL1_MODE_ONCE);      // start reading again
 }
 
-bool ak8963Detect(magDev_t *mag)
+static bool ak8963Init(void)
 {
-    uint8_t sig = 0;
+    uint8_t calibration[3];
+    uint8_t status;
 
-#if defined(USE_SPI) && defined(AK8963_SPI_INSTANCE)
-    spiBusSetInstance(&mag->bus, AK8963_SPI_INSTANCE);
-    mag->bus.busdev_u.spi.csnPin = mag->bus.busdev_u.spi.csnPin == IO_NONE ? IOGetByTag(IO_TAG(AK8963_CS_PIN)) : mag->bus.busdev_u.spi.csnPin;
+    ak8963WriteRegister(busdev, AK8963_MAG_REG_CNTL1, CNTL1_MODE_POWER_DOWN);       // power down before entering fuse mode
+    ak8963WriteRegister(busdev, AK8963_MAG_REG_CNTL1, CNTL1_MODE_FUSE_ROM);         // Enter Fuse ROM access mode
+    ak8963ReadRegisterBuffer(busdev, AK8963_MAG_REG_ASAX, sizeof(calibration), calibration); // Read the x-, y-, and z-axis calibration values
+
+    magGain[X] = ((((float)(int8_t)calibration[X] - 128) / 256) + 1) * 30;
```

Comment:
IMO calibration should be unsigned, it makes little sense with signed.
https://github.com/betaflight/betaflight/pull/3215#pullrequestreview-46061008